core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/pipelines/MaxSpeedEnvelope.kt

```diff
@@ -29,10 +47,16 @@ object MaxSpeedEnvelope {
      * Generate braking curves overlay everywhere the mrsp decrease (increase backwards) with a
      * discontinuity
      */
-    private fun addBrakingCurves(context: EnvelopeSimContext, mrsp: Envelope): Envelope {
-        val builder = OverlayEnvelopeBuilder.backward(mrsp)
-        val cursor = EnvelopeCursor.backward(mrsp)
-        var lastPosition = mrsp.endPos
+    private fun addBrakingCurves(
+        etcsSimulator: ETCSBrakingSimulator,
+        context: EnvelopeSimContext,
+        mrsp: Envelope
+    ): Envelope {
+        var envelope = mrsp
+        envelope = addETCSBrakingCurves(etcsSimulator, context, envelope)
+        val builder = OverlayEnvelopeBuilder.backward(envelope)
+        val cursor = EnvelopeCursor.backward(envelope)
+        var lastPosition = envelope.endPos
         while (cursor.findPartTransition(MaxSpeedEnvelope::increase)) {
```

Comment:
This one is a little different, adding a deceleration curve ending at a discontinuity may erase other discontinuities. If we pre-compute the discontinuities we'd need to filter them and it's not necessarily that easy to do, the current version feels more robust.

Not processing it strictly backwards means that we may compute unnecessary deceleration curves that would be later "absorbed" by a different one. It's not ideal but not really a big deal either.

Reply:
Functionally: As we can now have different shape of braking curves, we probably need to process them all and make sure to really use the minimum of all curves.
Ex: I imagine a quite "flat" ETCS braking curve ending late in the path, and in the middle of it (so ending sooner in the path) a steep const curve, we probably need to apply all curves. (this is a side attention point)

As for the code mutualisation: it may make sense to process curves one by one (const, ETCS LoA, ETCS EoA, ETCS SvL) and then apply the min for each of them.
If this is a good idea, it may also be done later.